crates/geoblock/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+//! Middleware which adds geo-location IP blocking.
+//!
+//! Note: this middleware requires you to use
+//! [Router::into_make_service_with_connect_info](https://docs.rs/axum/latest/axum/struct.Router.html#method.into_make_service_with_connect_info)
+//! to run your app otherwise it will fail at runtime.
+//!
+//! See [Router::into_make_service_with_connect_info](https://docs.rs/axum/latest/axum/struct.Router.html#method.into_make_service_with_connect_info) for more details.
+pub use geoip;
+use {
+    axum_client_ip::InsecureClientIp,
+    geoip::GeoIpResolver,
+    http_body::Body,
+    hyper::{Request, Response, StatusCode},
+    pin_project::pin_project,
+    std::{
+        future::Future,
+        net::IpAddr,
+        pin::Pin,
+        sync::Arc,
+        task::{Context, Poll},
+    },
+    thiserror::Error,
+    tower::Service,
+    tower_layer::Layer,
+    tracing::{error, info},
+};
+
+#[cfg(test)]
+mod tests;
+
+/// Values used to configure the middleware behavior when country information
+/// could not be retrieved.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum BlockingPolicy {
+    Block,
+    AllowMissingCountry,
+    AllowExtractFailure,
+    AllowAll,
+}
+
+#[derive(Debug, Error)]
+enum GeoBlockError {
+    #[error("Country is blocked")]
+    Blocked,
+
+    #[error("Unable to extract IP address")]
+    UnableToExtractIPAddress,
+
+    #[error("Unable to extract geo data from IP address")]
+    UnableToExtractGeoData,
+
+    #[error("Country could not be found in database")]
+    CountryNotFound,
+}
+
+/// Layer that applies the GeoBlock middleware which blocks requests base on IP
+/// geo-location.
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct GeoBlockLayer<R>
+where
+    R: GeoIpResolver,
+{
+    blocked_countries: Vec<Arc<str>>,
```

Comment:
These don't need to be `Arc<str>`. Regular `String`s should do fine. If we want to have a cheaper clone for the layer and the service (do we though?), it'd be better to use `Arc<Vec<String>>` than `Vec<Arc<str>>`, since the vec is always immutable.